
# HG changeset patch
# User welpy-cw <h.w.forms@arcor.de>
# Date 1699019553 -3600
# Node ID a3c504ca560a4e619c63c620b952c3dd4e26163d
# Parent  423d06bb9efea783e181d07f3b5da82dde585520
Bug 1862232 - Add a special "Reply to List" button to unified toolbar. r=freaktechnik


Checking if "Reply to List" should be enabled doesn't work until the headers of the current
message have been processed. This implements class ReplyListButton that listens for the "MsgLoaded"
event to set its state.

Differential Revision: https://phabricator.services.mozilla.com/D192471

diff --git a/mail/base/content/mailCommon.js b/mail/base/content/mailCommon.js
--- a/mail/base/content/mailCommon.js
+++ b/mail/base/content/mailCommon.js
@@ -546,17 +546,20 @@ var commandController = {
           ConversationOpener.isMessageIndexed(
             gDBView.hdrForFirstSelectedMessage
           )
         );
       case "cmd_replylist":
         if (hasIdentities && numSelectedMessages == 1) {
           const aboutMessage =
             document.getElementById("messageBrowser")?.contentWindow || window;
-          return aboutMessage?.currentHeaderData?.["list-post"];
+          return (
+            gDBView.URIForFirstSelectedMessage == aboutMessage?.gMessageURI &&
+            aboutMessage.currentHeaderData?.["list-post"]
+          );
         }
         return false;
       case "cmd_viewPageSource":
       case "cmd_saveAsTemplate":
         return numSelectedMessages == 1;
       case "cmd_reply":
       case "cmd_replySender":
       case "cmd_replyall":
diff --git a/mail/components/unifiedtoolbar/content/items/reply-list-button.mjs b/mail/components/unifiedtoolbar/content/items/reply-list-button.mjs
new file mode 100644
--- /dev/null
+++ b/mail/components/unifiedtoolbar/content/items/reply-list-button.mjs
@@ -0,0 +1,15 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { MailTabButton } from "chrome://messenger/content/unifiedtoolbar/mail-tab-button.mjs";
+
+/**
+ * Unified toolbar button for replying to a mailing list..
+ */
+class ReplyListButton extends MailTabButton {
+  observedAboutMessageEvents = ["load", "MsgLoaded"];
+}
+customElements.define("reply-list-button", ReplyListButton, {
+  extends: "button",
+});
diff --git a/mail/components/unifiedtoolbar/content/unifiedToolbarCustomizableItems.inc.xhtml b/mail/components/unifiedtoolbar/content/unifiedToolbarCustomizableItems.inc.xhtml
--- a/mail/components/unifiedtoolbar/content/unifiedToolbarCustomizableItems.inc.xhtml
+++ b/mail/components/unifiedtoolbar/content/unifiedToolbarCustomizableItems.inc.xhtml
@@ -77,17 +77,17 @@
   <button is="mail-tab-button"
           command="cmd_replyall"
           label-id="toolbar-reply-all-label"
           data-l10n-id="toolbar-reply-all"></button>
 </html:template>
 
 <html:template id="replyToListTemplate"
                xmlns="http://www.w3.org/1999/xhtml">
-  <button is="mail-tab-button"
+  <button is="reply-list-button"
           command="cmd_replylist"
           label-id="toolbar-reply-to-list-label"
           data-l10n-id="toolbar-reply-to-list"></button>
 </html:template>
 
 <html:template id="redirectTemplate"
                xmlns="http://www.w3.org/1999/xhtml">
   <button is="mail-tab-button"
diff --git a/mail/components/unifiedtoolbar/jar.mn b/mail/components/unifiedtoolbar/jar.mn
--- a/mail/components/unifiedtoolbar/jar.mn
+++ b/mail/components/unifiedtoolbar/jar.mn
@@ -15,15 +15,16 @@ messenger.jar:
     content/messenger/unifiedtoolbar/global-search-bar.mjs             (content/items/global-search-bar.mjs)
     content/messenger/unifiedtoolbar/mail-go-button.mjs                (content/items/mail-go-button.mjs)
     content/messenger/unifiedtoolbar/quick-filter-bar-toggle.mjs       (content/items/quick-filter-bar-toggle.mjs)
     content/messenger/unifiedtoolbar/space-button.mjs                  (content/items/space-button.mjs)
     content/messenger/unifiedtoolbar/view-picker-button.mjs            (content/items/view-picker-button.mjs)
     content/messenger/unifiedtoolbar/extension-action-button.mjs       (content/extension-action-button.mjs)
     content/messenger/unifiedtoolbar/list-box-selection.mjs            (content/list-box-selection.mjs)
     content/messenger/unifiedtoolbar/mail-tab-button.mjs               (content/mail-tab-button.mjs)
+    content/messenger/unifiedtoolbar/reply-list-button.mjs             (content/items/reply-list-button.mjs)
     content/messenger/unifiedtoolbar/search-bar.mjs                    (content/search-bar.mjs)
     content/messenger/unifiedtoolbar/unified-toolbar.mjs               (content/unified-toolbar.mjs)
     content/messenger/unifiedtoolbar/unified-toolbar-button.mjs        (content/unified-toolbar-button.mjs)
     content/messenger/unifiedtoolbar/unified-toolbar-customization.mjs (content/unified-toolbar-customization.mjs)
     content/messenger/unifiedtoolbar/unified-toolbar-customization-pane.mjs (content/unified-toolbar-customization-pane.mjs)
     content/messenger/unifiedtoolbar/unified-toolbar-tab.mjs           (content/unified-toolbar-tab.mjs)
     content/messenger/unifiedtoolbar/unifiedToolbarWebextensions.css   (content/unifiedToolbarWebextensions.css)
diff --git a/mail/components/unifiedtoolbar/modules/CustomizableItemsDetails.mjs b/mail/components/unifiedtoolbar/modules/CustomizableItemsDetails.mjs
--- a/mail/components/unifiedtoolbar/modules/CustomizableItemsDetails.mjs
+++ b/mail/components/unifiedtoolbar/modules/CustomizableItemsDetails.mjs
@@ -163,17 +163,17 @@ export default [
     ],
   },
   {
     id: "reply-to-list",
     labelId: "toolbar-reply-to-list",
     spaces: ["mail"],
     templateId: "replyToListTemplate",
     requiredModules: [
-      "chrome://messenger/content/unifiedtoolbar/mail-tab-button.mjs",
+      "chrome://messenger/content/unifiedtoolbar/reply-list-button.mjs",
     ],
   },
   {
     id: "redirect",
     labelId: "toolbar-redirect",
     spaces: ["mail"],
     templateId: "redirectTemplate",
     requiredModules: [
